src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <errno.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+
+#define SD_LISTEN_FDS_START 3
+#define SOCKET_MODES 3
```

Comment:
GLib has this lovely `G_N_ELEMENTS` macro that it might be useful to define an equivalent to here.  Then you can do N_ELEMENTS (instance_types) and avoid repeating the 3 here.